toolkit/tools/internal/timestamp/timestamp_mgr.go

```diff
@@ -0,0 +1,365 @@
+func StopEventByPath(path string) (ts *TimeStamp, err error) {
+	if err = ensureManagerExists(); err != nil {
+		return &TimeStamp{}, err
+	}
+
+	components := strings.Split(path, pathSeparator)
+	if components[0] != timestampMgr.root.Name {
+		err = fmt.Errorf("Timestamp root mismatch ('%s', expected '%s')", components[0], timestampMgr.root.Name)
+		return
+	}
+	ts, err = getTimeStampFromPath(timestampMgr.root, components, 1)
+	if err != nil {
+		return &TimeStamp{}, err
+	}
+
+	timestampMgr.submitEvent(EventStop, ts)
+	return
+}
+
+// Submit a recorded event to event queue to be processed
+func (mgr *TimeStampManager) submitEvent(eventType EventType, ts *TimeStamp) {
+	mgr.EventQueue <- &TimeStampRecord{EventType: eventType, TimeStamp: ts, time: time.Now()}
+}
+
+// Process each event submitted to the queue by updating in-memory data structure and recording to file
+func (mgr *TimeStampManager) processEventsInQueue() {
+	for event := range mgr.EventQueue {
+		mgr.updateRead(event)
+		mgr.writeToFile(event)
+	}
+	mgr.eventProcessorFinished <- true
+}
+
+// Append a timestamp record to file, and periodically flush to disk
+func (writeMgr *TimeStampWriteManager) writeToFile(record *TimeStampRecord) {
+	outputBytes, err := json.Marshal(record)
+	if err != nil {
+		logger.Log.Warnf("Failed to marshal timestamp record: %v", err)
+		return
+	}
+	writeMgr.writeBuffer = append(writeMgr.writeBuffer, outputBytes)
+
+	cooldownDuration := time.Duration(writeCooldownMilliseconds) * time.Millisecond
+	if writeMgr.lastWrite.Add(cooldownDuration).After(time.Now()) {
+		return
+	}
+
+	writeMgr.flush()
+	writeMgr.lastWrite = time.Now()
+}
+
+// Flush write buffer to disk
+func (writeMgr *TimeStampWriteManager) flush() {
+	for _, outputBytes := range writeMgr.writeBuffer {
+		_, err := writeMgr.fileDescriptor.WriteString(string(outputBytes) + "\n")
+		if err != nil {
+			logger.Log.Warnf("Failed to write timestamp record to file %s: %v", writeMgr.filePath, err)
+			continue
+		}
+	}
+	err := writeMgr.fileDescriptor.Sync()
+	if err != nil {
+		logger.Log.Warnf("Failed to save file %s to disk: %v", writeMgr.filePath, err)
+		return
+	}
+	writeMgr.writeBuffer = nil
+}
+
+// Update the timestamp tree for each event submitted to the event queue
+func (mgr *TimeStampManager) updateRead(record *TimeStampRecord) {
+	if record.TimeStamp == nil {
+		record.TimeStamp = mgr.lastVisited
+	}
+
+	switch record.EventType {
+	case EventStart:
+		ts := record.TimeStamp
+		ts.StartTime = &record.time
+		mgr.nodes[ts.ID] = ts
+		if record.parentTimestamp == nil {
+			if mgr.lastVisited == nil {
+				mgr.root = ts
+			} else {
+				mgr.lastVisited.addSubStep(ts)
+			}
+			mgr.lastVisited = ts
+		} else {
+			mgr.nodes[ts.ParentID].addSubStep(ts)
+			if mgr.lastVisited.ID == ts.ParentID {
+				mgr.lastVisited = ts
+			}
+		}
+	case EventStop:
+		ts := mgr.nodes[record.ID]
+		ts.complete(record.time)
+		if record.EventType == EventStop && ts.ID == mgr.lastVisited.ID {
```

Comment:
Not sure if you want to handle this or not, its kind of user error in a way. This can leave dangling substeps without a stop event. i.e. if we start a bunch of steps, then call `CompleteTiming()` without stopping the stub steps first we get an unbalanced set of events.

Reply:
There will be some incorrectness in the measurement whether or not we handle that. We can either have a check at the end and print out a warning if some start events do not have a corresponding stop events, or we can force stop all existing events when `CompleteTiming()` is called. I think we can go with either way.

Reply:
The original implementation did the second option, but I'm not sure its a big deal either way.